app/workers/batch_geocode_worker.rb

```diff
@@ -0,0 +1,29 @@
+# frozen_string_literal: true
+
+class BatchGeocodeWorker
+  include Sidekiq::Worker
+  sidekiq_options retry: 25
+
+  def perform(map_id)
+    map = Map.find_by(id: map_id)
+    return if map.nil?
+
+    locations_to_geocode = map.locations.geocoding_pending
+    return if locations_to_geocode.empty?
+
+    # Mapbox Geocoding limit is 600 per minute, we geocode max. 500 locations per batch
+    # @see https://docs.mapbox.com/api/search/geocoding/#geocoding-restrictions-and-limits
+    locations_to_geocode.find_in_batches(batch_size: 500).with_index do |group, batch|
+      group.each do |location|
+        location.geocode
+        location.save
+      end
+
+      sleep(60) unless batch.zero?
+    end
+
+    # If we still have pending locations at this stage, retry the worker
+    count_after = map.locations.geocoding_pending.count
+    raise "There are still #{count_after} pending locations to geocode" if count_after.positive?
```

Comment:
The retry with raising an error feels a bit hacky. Can't we just call `perform` again?

Reply:
Yes we can call `perform.new` again. The good thing about raising an error is that Sidekiq will apply a smarter retry logic to it (https://www.rubydoc.info/gems/sidekiq/Sidekiq/JobRetry). I think it's pretty standard to do it this way

Reply:
I see. It just doesn't seem right to retry even though the job worked correctly... So I think starting a new job would be better? We can also leave this for a later task/discussion.

Reply:
> It just doesn't seem right to retry even though the job worked correctly

Yes I mean if the job worked correctly, we won't retry it. The `count_after` will be 0 if all pending locations have been geocoded. So the retry happens actually only if the geocoding has failed for some reason

Reply:
Yes I know. But exceeding the batch size is not really an error, we know that it is going to happen before. It feels like abusing the error system and sidekiq. But I think it's fine for now.

Reply:
> But exceeding the batch size is not really an error

No no, it's not about the batch size. There is no batch size limit, it will loop until all the pending locations have been through. It's about geocoding error.

`location.geocode` is not configured to raise an error. So if we reach the quota limit (still, even if we are limiting to 500 per minute) or because of a timeout, it will silently put `latitude` and `longitude` of this location to `nil` and continue with the other locations. It doesn't raise an exception. So basically the `count_after` line is to check if there was geocoding errors in the process. If yes, retry the job again.

Reply:
Ahhhh. Sorry, misunderstood the code. In this case yes, all good!